frontend/src/Language/Granule/Checker/Checker.hs

```diff
@@ -1599,8 +1599,18 @@ synthExpr defs gam pol e@(AppTy s _ rf e1 ty) = do
           debugM "derived drop tys:" (pretty typScheme)
           -- return this variable expression in place here
           freshenTySchemeForVar s rf name typScheme
-    _ -> throw NeedTypeSignature{ errLoc = getSpan e, errExpr = e }
 
+    expr -> do
+      -- Synth expr, expecting it to be a forall now
+      (retTy, gam', subst, elab) <- synthExpr defs gam pol expr
+      case retTy of
```

Comment:
Just a question/remark (don't know the coding style!): could this case be joined into an irrefutable pattern in line 1605?

Reply:
Not really because it's not irrefutable- we need to be able to raise exceptions if the return type is of the wrong form.